block-layer/codecs/dag-jose.md

```diff
@@ -0,0 +1,126 @@
+
+```ipldsch
+type GeneralSignature struct {
+  header optional {String:Any}
+  protected optional String
+  signature String
+}
+
+type GeneralJWS struct {
+  payload String
+  signatures [GeneralSignature]
+}
+
+type GeneralRecipient struct {
+  encrypted_key optional String
+  header optional {String:Any}
+}
+
+type GeneralJWE struct {
+  aad optional String
+  ciphertext String
+  iv optional String
+  protected optional String
+  recipients [GeneralRecipient]
+  tag optional String
+  unprotected optional {String:Any}
+}
+
+type GeneralJOSE union {
+  | GeneralJWS jws
+  | GeneralJWE jwe
+} representation kinded
+```
+
+##### Serializing the Decoded Representation
+
+When serializing a JOSE object from the `Decoded Representation` special care needs to be taken with the `payload` property as well as the `protected` properties. 
+
+###### Protected
+
+The `protected` property in JWE and JWS have the type `{String:Any}`. This means that it may include data with *Link Kind* and *Bytes Kind*. These should be converted into pure JSON in the same way as it's done in [DAG-JSON](./dag-json.md). However, the properties should **not** be sorted since that would cause any integrity check on the JOSE data to fail. Once in JSON format the `protected` property should be converted into `base64url` using the method described in the JOSE spec  (`BASE64URL(UTF8(data))`). 
+
+###### Payload
+
+The payload property of JWS can be of either `Bytes` or  `{String:Any}` types. If the former it's simply just encoded as `base64url`. If the latter, it should be encoded in the same manner as the `protected` property.
+
+Note that any change in the ordering of the properties of the payload at this point would cause potential validation of the JOSE object to fail. Good signature libraries will sort the payload before the signature is applied.
+
+#### Ordering
+
+Once the data has been converted to the `General Serialization`, codec implementors **MUST** use the same sorting algorithm as [DAG-JSON](./dag-json.md) to sort the data to ensure hashes consistently match for the same block data.
+
+## Additional information
+
+### Reccomended JOSE creation strategy
+
+When creating a JOSE object there are some suggested approaches of how to format the data that is being signed / encrypted / authenticated that will keep you out of trouble. The main thing to keep in mind is that signatures / data authentication could be invalidated if the order of the properties in the JOSE object changes. It's therefore a good idea to sort the properties before any signature / authentication is added. The best way to do this is simply to use the same strategy employed by [DAG-JSON](./dag-json.md), which will also convert `Link` and `Bytes` to JSON representation. 
```

Comment:
This is very lose advice, it doesn't pin implementers to any particular strategy, what's the reason for not being more strict about strategy here? We're finding the lose edges across our various codecs create awkward windows for same-data-different-hash among other problems. The more "must" and less "should"/"could"/"may" the better unless there's a good reason to be flexible. Are we trying to include any arbitrary JOSE document in this spec? Would it be better to have a separate JOSE codec and have DAG-JOSE full of "musts" where it's very restrictive about how you order and lay things and what's possible and what's not.